orca/orca.go

```diff
@@ -0,0 +1,175 @@
+		r := &CallMetricRecorder{recorder: newMetricRecorder()}
+		ctxWithRecorder := setCallMetricRecorder(ctx, r)
+		resp, err := handler(ctxWithRecorder, req)
+		if err2 := r.recorder.setTrailerMetadata(ctx); err2 != nil {
+			logger.Warning(err2)
+		}
+		return resp, err
+	}
+}
+
+// wrappedStream wraps the grpc.ServerStream received by the streaming
+// interceptor. Overrides only the Context() method to return a context which
+// contains a reference to the CallMetricRecorder corresponding to this stream.
+type wrappedStream struct {
+	grpc.ServerStream
+	ctx context.Context
+}
+
+func (w *wrappedStream) Context() context.Context {
+	return w.ctx
+}
+
+// EnableMetricsReportingForStreamingRPCs is similar to
+// EnableReportingForUnaryRPCs, except that is it to be used for streaming RPCs.
+func EnableMetricsReportingForStreamingRPCs() grpc.StreamServerInterceptor {
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		r := &CallMetricRecorder{recorder: newMetricRecorder()}
+		ctxWithRecorder := setCallMetricRecorder(ss.Context(), r)
+		err := handler(srv, &wrappedStream{
+			ServerStream: ss,
+			ctx:          ctxWithRecorder,
+		})
+		if err2 := r.recorder.setTrailerMetadata(ss.Context()); err2 != nil {
+			logger.Warning(err2)
+		}
+		return err
+	}
+}
+
+// minOutOfBandReportingInterval is the absolute minimum value supported for
+// out-of-band metrics reporting.
+const minOutOfBandReportingInterval = 30 * time.Second
+
+// OutOfBandMetricsReportingOptions contains options to configure out-of-metrics
+// metrics reporting from server applications.
+type OutOfBandMetricsReportingOptions struct {
+	// MinReportingInterval sets the lower bound for how often out-of-band
+	// metrics are reported on the streaming RPC initiated by the client.  If
+	// unspecified or less than the default value of 30s, the default is used.
+	// Clients may request a higher value as part of the StreamCoreMetrics
+	// streaming RPC.
+	MinReportingInterval time.Duration
+}
+
+// EnableOutOfBandMetricsReporting registers an implementation of the
+// [OpenRcaService] on the provider service registrar [grpc.ServiceRegistrar].
+//
+// [OpenRcaService]: https://github.com/cncf/xds/blob/main/xds/service/orca/v3/orca.proto
+func EnableOutOfBandMetricsReporting(s grpc.ServiceRegistrar, opts OutOfBandMetricsReportingOptions) error {
+	if opts.MinReportingInterval < minOutOfBandReportingInterval {
+		opts.MinReportingInterval = minOutOfBandReportingInterval
+	}
+	return enableOutOfBandMetricsReporting(s, opts)
+}
+
+// EnableOutOfBandMetricsReportingForTesting registers an implementation of the
+// [OpenRcaService] on the provider service registrar [grpc.ServiceRegistrar].
+// Similar to EnableOutOfBandMetricsReporting, except no validation is done on
+// the provided options.
+//
+// [OpenRcaService]: https://github.com/cncf/xds/blob/main/xds/service/orca/v3/orca.proto
+//
+// Testing Only
```

Comment:
Could this be moved to `internal/orca` instead?  Or do we think users will want this (in which case, can it move to `orca/testing`?)?

Reply:
Got rid of this totally. Moved the reporting interval `var` to `internal` package.